ocrd_cis/ocropy/clip.py

```diff
@@ -39,6 +37,113 @@ def executable(self):
+                    continue
+                shape = prep(shapes[i])
+                neighbours = [(regionj, maskj) for shapej, regionj, maskj
+                              in zip(shapes[:i] + shapes[i+1:], regions[:i] + regions[i+1:], masks[:i] + masks[i+1:])
+                              if shape.intersects(shapej)]
+                if neighbours:
+                    segment_region_file_id = f"{output_file_id}_{region.id}"
+                    ret.append(self.process_segment(
+                        region, masks[i], polygons[i], neighbours, background_image,
+                        page_image, page_coords, page_bin, page_id, segment_region_file_id))
+                continue
+            # level == 'line':
+            lines = region.get_TextLine()
+            if not lines:
+                self.logger.warning(f'Page "{page_id}" region "{region.id}" contains no text lines')
+                continue
+            region_image, region_coords = self.workspace.image_from_segment(
+                region, page_image, page_coords, feature_selector='binarized')
+            background_image = Image.new(region_image.mode, region_image.size, background)
+            region_array = pil2array(region_image)
+            region_bin = np.array(region_array <= midrange(region_array), np.uint8)
+            # in absolute coordinates merely for comparison/intersection
+            shapes = [Polygon(polygon_from_points(line.get_Coords().points)) for line in lines]
+            # in relative coordinates for mask/cropping
+            polygons = [coordinates_of_segment(line, region_image, region_coords) for line in lines]
+            masks = [pil2array(polygon_mask(region_image, polygon)).astype(np.uint8) for polygon in polygons]
+            for j, line in enumerate(lines):
+                if line.get_AlternativeImage():
+                    # FIXME: This should probably be an exception (bad workflow configuration).
+                    self.logger.warning(
+                        f'Page "{page_id}" region "{region.id}" line "{line.id}" already contains image '
+                        f'data: skipping')
+                    continue
+                shape = prep(shapes[j])
+                neighbours = [(linej, maskj) for shapej, linej, maskj
+                              in zip(shapes[:j] + shapes[j+1:], lines[:j] + lines[j+1:], masks[:j] + masks[j+1:])
+                              if shape.intersects(shapej)]
+                if neighbours:
+                    segment_line_file_id = f"{output_file_id}_{region.id}_{line.id}"
+                    ret.append(self.process_segment(
+                        line, masks[j], polygons[j], neighbours, background_image,
+                        region_image, region_coords, region_bin, page_id, segment_line_file_id))
+        return ret
+
+    # TODO: remove when `process_page_pcgts` is validated to be correct
     def process(self):
```

Comment:
I got a bit confused because there are changes now in `process`, which we want to get rid of. After `process_segment` has been changed,  `process` won't work anyway.

Otherwise perfectly fine – thank you!

Reply:
Since I am not comfortable with the processor implementation codebase yet, I refactored the `process()` method to have fewer lines and refined some places. Then I implemented the `process_page_pcgts()` step by step and made sure I did not mess up badly. That is why you also see process() updated. I will remove the `process()` method since it is harder to review. Sorry for the confusion.